lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp

```diff
@@ -208,4 +208,301 @@ void AmfNasStateConverter::proto_to_ue_m5gmm_context(
   // Actual implementation logic will be added as part of upcoming pr
 }
 
+void AmfNasStateConverter::smf_proc_data_to_proto(
+    const smf_proc_data_t* state_smf_proc_data,
+    magma::lte::oai::Smf_Proc_Data* smf_proc_data_proto) {
+  smf_proc_data_proto->set_pdu_session_id(state_smf_proc_data->pdu_session_id);
+  smf_proc_data_proto->set_pti(state_smf_proc_data->pti);
+  smf_proc_data_proto->set_message_type(
+      static_cast<uint32_t>(state_smf_proc_data->message_type));
+  smf_proc_data_proto->set_max_uplink(state_smf_proc_data->max_uplink);
+  smf_proc_data_proto->set_max_downlink(state_smf_proc_data->max_downlink);
+  smf_proc_data_proto->set_pdu_session_type(
+      (magma::lte::oai::M5GPduSessionType)
+          state_smf_proc_data->pdu_session_type);
+  smf_proc_data_proto->set_ssc_mode(state_smf_proc_data->ssc_mode);
+}
+void AmfNasStateConverter::proto_to_smf_proc_data(
+    const magma::lte::oai::Smf_Proc_Data& smf_proc_data_proto,
+    smf_proc_data_t* state_smf_proc_data) {
+  state_smf_proc_data->pdu_session_id = smf_proc_data_proto.pdu_session_id();
+  state_smf_proc_data->pti            = smf_proc_data_proto.pti();
+  state_smf_proc_data->message_type =
+      static_cast<M5GMessageType>(smf_proc_data_proto.message_type());
+  state_smf_proc_data->max_uplink   = smf_proc_data_proto.max_uplink();
+  state_smf_proc_data->max_downlink = smf_proc_data_proto.max_downlink();
+  state_smf_proc_data->pdu_session_type =
+      static_cast<M5GPduSessionType>(smf_proc_data_proto.pdu_session_type());
+  state_smf_proc_data->ssc_mode = smf_proc_data_proto.ssc_mode();
+}
+
+void AmfNasStateConverter::s_nssai_to_proto(
+    const s_nssai_t* state_s_nssai, magma::lte::oai::SNssai* snassi_proto) {
+  snassi_proto->set_sst(state_s_nssai->sst);
+  snassi_proto->set_sd(*(uint32_t*) &state_s_nssai->sd);
```

Comment:
No need to convert to pointer and then dereference.

```suggestion
  snassi_proto->set_sd(state_s_nssai->sd);
```

Reply:
Here "sd" is a 3-byte raw buffer (uint8_t) but we are storing it as uint32 in protobuff. So this conversion is required. We thought of a couple of approaches and implemented this.
The other approach would be to store sd as "bytes" type in protobuff. But this approach will need conversion to char and then storing it and converting it back while retrieving.
Please suggest what would be the best approach to this.

Other approach pseudo code(sd is of type "bytes"):
void AmfNasStateConverter::sd_to_chars(const uint8_t* sd, char* sd_array) {
  sd_array[0] = static_cast<char>(sd[0] + ASCII_ZERO);
  sd_array[1] = static_cast<char>(sd[1] + ASCII_ZERO);
  sd_array[2] = static_cast<char>(sd[2] + ASCII_ZERO);
}

void AmfNasStateConverter::chars_to_sd(const char* sd_array, uint8_t* sd) {
  sd[0] = static_cast<int>(sd_array[0]) - ASCII_ZERO;
  sd[1] = static_cast<int>(sd_array[1]) - ASCII_ZERO;
  sd[2] = static_cast<int>(sd_array[2]) - ASCII_ZERO;
}

Reply:
@shashidhar-patil , you can use the "bytes" type in protobuf and pass the uint8_t array as char array:

State to proto
```
snassi_proto->set_sd((char*) state_s_nssai->sd, SD_LENGTH);
```

Proto to state
```
memcpy((void*)state_s_nssai->sd, (void*)snassi_proto.sd().c_str(), SD_LENGTH);
```

Note that you do not need to convert it into an ASCII value unless you want to print it as text.

Reply:
Thankyou @ssanadhya. Addressed this review comment.